sycl/doc/extensions/proposed/sycl_ext_oneapi_graph.asciidoc

```diff
@@ -0,0 +1,1050 @@
+`queue::submit(command_graph<graph_state::executable>)`.
+
+[source, c++]
+----
+  queue q{default_selector{}};
+
+  // New object representing graph of command-groups
+  ext::oneapi::experimental::command_graph graph;
+  {
+    buffer<T> bufferA{dataA.data(), range<1>{elements}};
+    buffer<T> bufferB{dataB.data(), range<1>{elements}};
+    buffer<T> bufferC{dataC.data(), range<1>{elements}};
+
+    // `q` will be put in the recording state where commands are recorded to
+    // `graph` rather than submitted for execution immediately.
+    q.begin_recording(graph);
+
+    // Record commands to `graph` with the following topology.
+    //
+    //      increment_kernel
+    //       /         \
+    //   A->/        A->\
+    //     /             \
+    //   add_kernel  subtract_kernel
+    //     \             /
+    //   B->\        C->/
+    //       \         /
+    //     decrement_kernel
+
+    q.submit([&](handler &cgh) {
+      auto pData = bufferA.get_access<access::mode::read_write>(cgh);
+      cgh.parallel_for<increment_kernel>(range<1>(elements),
+                                         [=](item<1> id) { pData[id]++; });
+    });
+
+    q.submit([&](handler &cgh) {
+      auto pData1 = bufferA.get_access<access::mode::read>(cgh);
+      auto pData2 = bufferB.get_access<access::mode::read_write>(cgh);
+      cgh.parallel_for<add_kernel>(range<1>(elements),
+                                   [=](item<1> id) { pData2[id] += pData1[id]; });
+    });
+
+    q.submit([&](handler &cgh) {
+      auto pData1 = bufferA.get_access<access::mode::read>(cgh);
+      auto pData2 = bufferC.get_access<access::mode::read_write>(cgh);
+      cgh.parallel_for<subtract_kernel>(
+          range<1>(elements), [=](item<1> id) { pData2[id] -= pData1[id]; });
+    });
+
+    q.submit([&](handler &cgh) {
+      auto pData1 = bufferB.get_access<access::mode::read_write>(cgh);
+      auto pData2 = bufferC.get_access<access::mode::read_write>(cgh);
+      cgh.parallel_for<decrement_kernel>(range<1>(elements), [=](item<1> id) {
+        pData1[id]--;
+        pData2[id]--;
+      });
+    });
+
+    // queue will be returned to the executing state where commands are
+    // submitted immediately for extension.
+    q.end_recording();
+  }
+
+  // Finalize the modifiable graph to create an executable graph that can be
+  // submitted for execution.
+  auto exec_graph = graph.finalize(q.get_context());
+
+  // Execute graph
+  q.submit(exec_graph);
+----
+
+== Issues
+
+=== Multi Device Graph
+
+Allow an executable graph to contain nodes targeting different devices.
+
+**Outcome:** Under consideration
+
+=== Record & Replay: Mark Internal Memory
+
+When a graph is created by recording a queue there is no way to tag memory
+objects internal to the graph, which would enable optimizations on the internal
+memory. Do we need an interface record & replay can use to identify buffers and
+USM allocations not used outside of the graph?
```

Comment:
Would such internal memory be per graph or shared between different submitted graphs (different graph instances)? Or do we need to handle both?